core/promise_ring.rs

```diff
@@ -0,0 +1,73 @@
+use crate::PromiseId;
+use std::collections::btree_map::BTreeMap;
+
+type PromiseResolver = v8::Global<v8::PromiseResolver>;
+const RING_SIZE: usize = 4 * 1024;
+
+// NOTE: could generalize to a "RingMap"
+pub struct PromiseRing {
+  len: u32,
+  cursor: u32,
+  // TODO(@AaronO): Maybe optimize by tracking existence with a bitset
+  // would reduce mem usage from 2*usize*RING_SIZE to (1+1/8)*usize*RING_SIZE
+  // with the current approach the Option tag "wastes" 32kb per isolate
+  ring: Vec<Option<PromiseResolver>>,
+  map: BTreeMap<PromiseId, PromiseResolver>,
+}
+
+impl PromiseRing {
+  pub(crate) fn new() -> Self {
+    Self {
+      len: 0,
+      cursor: 0,
+      ring: vec![None; RING_SIZE],
+      map: BTreeMap::new(),
+    }
+  }
+
+  // TODO(@AaronO): decide if this is useful (but it should match length of futures stream so ...)
+  // pub(crate) fn len(&self) -> u32 {
+  //   self.len
+  // }
+
+  pub(crate) fn has(&self, id: PromiseId) -> bool {
+    self.ring.get(Self::ring_idx(id)).is_some() || self.map.contains_key(&id)
```

Comment:
Can the ring and map be desynced such that the double check is necessary?

Reply:
Only checking either of them should be sufficent (based on if the id is in the range of the ring or not), but it's unimportant to optimize since this should be called infrequently for ref/unref ops

Reply:
I'm not sure this is right. If a promise resolver is moved to the map, it's because a new resolver with a different id took its slot. If it is then removed from the map but the resolver that took its slot hasn't been removed yet, this method would return `true`.

Also, `self.ring.get(idx)` has type `Option<&Option<PromiseResolver>>`, and the outer option is some if the index is a valid index into the slice – which it's guaranteed to be by `Self::ring_idx`. So `self.ring.get(Self::ring_idx(id)).is_some()` always returns true, even if cursor is 0.
```suggestion
    self.ring[Self::ring_idx(id)].is_some() || self.map.contains_key(&id)
```